File: Sorting-Algorithms/BucketSort/BucketSort.c
#include <stdio.h>
#include <stdlib.h>

void insertion_sort(int arr[], int n) {
    for (int i = 1; i < n; i++) {
        int key = arr[i], j = i - 1;

        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

void bucket_sort(int arr[], int n) {
    int max = arr[0];

    for (int i = 1; i < n; i++)
        if (arr[i] > max) max = arr[i];

    int bucket[n][n], count[n];

    for (int i = 0; i < n; i++)
        count[i] = 0;

    // distribute elements
    for (int i = 0; i < n; i++) {
        int index = arr[i] * n / (max + 1);
        bucket[index][count[index]++] = arr[i];
    }

    // sort buckets
    for (int i = 0; i < n; i++)
        insertion_sort(bucket[i], count[i]);

    // merge
    int k = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < count[i]; j++)
            arr[k++] = bucket[i][j];
    }
}

void print(int arr[], int n) {
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

int main() {
    int arr[100], n;

    printf("Enter n: ");
    scanf("%d", &n);

    for (int i = 0; i < n; i++)
        scanf("%d", &arr[i]);

    printf("Original: ");
    print(arr, n);

    bucket_sort(arr, n);

    printf("Sorted: ");
    print(arr, n);

    return 0;
}

File: Sorting-Algorithms/BucketSort/BucketSort.py
def bucket_sort(arr):
    if len(arr) == 0:
        return arr

    max_val = max(arr)
    bucket_count = len(arr)

    buckets = [[] for _ in range(bucket_count)]

    # distribute elements into buckets
    for num in arr:
        index = int(num * bucket_count / (max_val + 1))
        buckets[index].append(num)

    # sort individual buckets
    for bucket in buckets:
        bucket.sort()

    # merge buckets
    k = 0
    for bucket in buckets:
        for num in bucket:
            arr[k] = num
            k += 1


def main():
    while True:
        print("\n===== MENU =====")
        print("1. User Input")
        print("2. Test Cases")
        print("3. Exit")

        choice = int(input("Enter choice: "))

        if choice == 1:
            n = int(input("Enter number of elements: "))
            arr = [int(input(f"Enter element {i+1}: ")) for i in range(n)]

            print("Original:", arr)
            bucket_sort(arr)
            print("Sorted:", arr)

        elif choice == 2:
            test_cases = [
                [1, 2, 3, 4, 5],
                [10, 20, 30, 40],
                [3, 1, 4, 2, 5],
                [8, 3, 7, 2, 6],
                [5, 4, 3, 2, 1],
                [9, 7, 5, 3, 1]
            ]

            for i in range(6):
                print(f"{i+1}. Test Case {i+1}")

            t = int(input("Enter choice: "))

            if 1 <= t <= 6:
                arr = test_cases[t - 1].copy()
                print("Original:", arr)
                bucket_sort(arr)
                print("Sorted:", arr)
            else:
                print("Invalid choice!")

        elif choice == 3:
            break

main()